File: frontend/src/lib/api/studio/mind-maps.ts
/**
 * Mind Maps API
 * Educational Note: Handles AI-generated mind maps for visualizing content structure.
 */

import axios from 'axios';
import { API_BASE_URL } from '../client';
import type { JobStatus } from './index';
import { createLogger } from '@/lib/logger';

const log = createLogger('studio-mind-maps-api');

/**
 * Mind map node from Claude
 */
export interface MindMapNode {
  id: string;
  label: string;
  parent_id: string | null;
  node_type: 'root' | 'category' | 'leaf';
  description: string;
}

/**
 * Mind map job record from the API
 */
export interface MindMapJob {
  id: string;
  source_id: string;
  source_name: string;
  direction: string;
  status: JobStatus;
  progress: string;
  error: string | null;
  nodes: MindMapNode[];
  topic_summary: string | null;
  node_count: number;
  generation_time_seconds: number | null;
  // Edit lineage
  parent_job_id: string | null;
  edit_instructions: string | null;
  created_at: string;
  started_at: string | null;
  completed_at: string | null;
}

/**
 * Response from starting mind map generation
 */
export interface StartMindMapResponse {
  success: boolean;
  job_id?: string;
  message?: string;
  source_name?: string;
  error?: string;
}

/**
 * Response from getting mind map job status
 */
export interface MindMapJobStatusResponse {
  success: boolean;
  job?: MindMapJob;
  error?: string;
}

/**
 * Response from listing mind map jobs
 */
export interface ListMindMapJobsResponse {
  success: boolean;
  jobs: MindMapJob[];
  count: number;
  error?: string;
}

/**
 * Mind Maps API
 */
export const mindMapsAPI = {
  /**
   * Start mind map generation
   */
  async startGeneration(
    projectId: string,
    sourceId: string,
    direction?: string,
    parentJobId?: string,
    editInstructions?: string
  ): Promise<StartMindMapResponse> {
    try {
      const body: Record<string, unknown> = {
        source_id: sourceId,
        direction: direction || 'Create a mind map covering the key concepts and their relationships.',
      };
      if (parentJobId) body.parent_job_id = parentJobId;
      if (editInstructions) body.edit_instructions = editInstructions;

      const response = await axios.post(
        `${API_BASE_URL}/projects/${projectId}/studio/mind-map`,
        body
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response) {
        return error.response.data;
      }
      log.error({ err: error }, 'failed to start mind map generation');
      throw error;
    }
  },

  /**
   * Get the status of a mind map job
   */
  async getJobStatus(projectId: string, jobId: string): Promise<MindMapJobStatusResponse> {
    try {
      const response = await axios.get(
        `${API_BASE_URL}/projects/${projectId}/studio/mind-map-jobs/${jobId}`
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response) {
        return error.response.data;
      }
      log.error({ err: error }, 'failed to get mind map job status');
      throw error;
    }
  },

  /**
   * List all mind map jobs for a project
   */
  async listJobs(projectId: string, sourceId?: string): Promise<ListMindMapJobsResponse> {
    try {
      const params = sourceId ? { source_id: sourceId } : {};
      const response = await axios.get(
        `${API_BASE_URL}/projects/${projectId}/studio/mind-map-jobs`,
        { params }
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response) {
        return error.response.data;
      }
      log.error({ err: error }, 'failed to list mind map jobs');
      throw error;
    }
  },

  /**
   * Delete a mind map job
   */
  async deleteJob(projectId: string, jobId: string): Promise<{ success: boolean; error?: string }> {
    try {
      const response = await axios.delete(
        `${API_BASE_URL}/projects/${projectId}/studio/mind-map-jobs/${jobId}`
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response) {
        return error.response.data;
      }
      log.error({ err: error }, 'failed to delete mind map job');
      throw error;
    }
  },

  /**
   * Poll mind map job status until complete or error
   */
  async pollJobStatus(
    projectId: string,
    jobId: string,
    onProgress?: (job: MindMapJob) => void,
    intervalMs: number = 2000,
    maxAttempts: number = 60
  ): Promise<MindMapJob> {
    let attempts = 0;
    let currentInterval = intervalMs;

    while (attempts < maxAttempts) {
      const response = await this.getJobStatus(projectId, jobId);

      if (!response.success || !response.job) {
        throw new Error(response.error || 'Failed to get job status');
      }

      const job = response.job;

      if (onProgress) {
        onProgress(job);
      }

      if (job.status === 'ready' || job.status === 'error') {
        return job;
      }

      await new Promise((resolve) => setTimeout(resolve, currentInterval));

      attempts++;

      if (attempts > 5 && currentInterval < 5000) {
        currentInterval = Math.min(currentInterval * 1.2, 5000);
      }
    }

    throw new Error('Mind map generation timed out');
  },
};

File: backend/app/api/studio/mind_maps.py
"""
Mind Map endpoints - AI-generated concept hierarchies.

Educational Note: Mind maps demonstrate hierarchical content extraction:
1. Claude analyzes source content
2. Identifies main topics and subtopics
3. Creates a tree structure of concepts
4. Returns structured JSON for visual rendering

Node Structure Pattern:
- Each node has: id, label, children[]
- Root node is the main topic
- Branches represent subtopics/details
- Frontend renders as interactive diagram

Visualization:
- Frontend uses libraries like react-flow or d3
- Nodes are positioned using layout algorithms
- Users can expand/collapse branches
- Colors indicate topic categories

Routes:
- POST /projects/<id>/studio/mind-map           - Start generation
- GET  /projects/<id>/studio/mind-map-jobs/<id> - Job status
- GET  /projects/<id>/studio/mind-map-jobs      - List jobs
"""
import json
import uuid
from flask import jsonify, request, current_app
from app.api.studio import studio_bp
from app.services.studio_services import studio_index_service
from app.services.studio_services.mind_map_service import mind_map_service
from app.services.source_services import source_index_service
from app.services.background_services.task_service import task_service
from app.services.auth import require_permission


@studio_bp.route('/projects/<project_id>/studio/mind-map', methods=['POST'])
@require_permission("studio", "mind_maps")
def generate_mind_map(project_id: str):
    """
    Start mind map generation or edit as a background task.

    Educational Note: Mind maps are generated from source content using
    Claude to create hierarchical node structures for visual concept mapping.
    Edits refine existing mind maps based on user instructions.

    Request Body:
        - source_id: UUID of the source to generate mind map from (required)
        - direction: Optional guidance for what to focus on
        - parent_job_id: UUID of the parent job to edit (optional, for edits)
        - edit_instructions: Instructions for editing the parent mind map (optional, for edits)

    Response:
        - success: Boolean
        - job_id: ID for polling status
        - message: Status message
    """
    try:
        data = request.get_json() or {}

        source_id = data.get('source_id')
        if not source_id:
            return jsonify({
                'success': False,
                'error': 'source_id is required'
            }), 400

        direction = data.get('direction', 'Create a mind map covering the key concepts and their relationships.')

        # Edit mode: load parent job's nodes as context for refinement
        parent_job_id = data.get('parent_job_id')
        edit_instructions = data.get('edit_instructions')
        previous_content = None

        # Reject edit requests without instructions — otherwise the service
        # falls through to a full regeneration with a misleading "Edited" badge.
        if parent_job_id and not edit_instructions:
            return jsonify({
                'success': False,
                'error': 'edit_instructions is required when parent_job_id is provided'
            }), 400

        if parent_job_id:
            # Clean up any previously failed edit jobs for this parent
            try:
                all_jobs = studio_index_service.list_mind_map_jobs(project_id)
                for job in all_jobs:
                    if (job.get("status") == "error"
                            and job.get("parent_job_id") == parent_job_id):
                        studio_index_service.delete_mind_map_job(project_id, job["id"])
            except Exception:
                pass  # Non-critical cleanup — don't block the new edit

            parent_job = studio_index_service.get_mind_map_job(project_id, parent_job_id)
            if not parent_job:
                return jsonify({
                    'success': False,
                    'error': 'Parent job not found'
                }), 404

            # Serialize previous nodes as JSON string for LLM context
            nodes = parent_job.get('nodes', [])
            if not nodes:
                return jsonify({
                    'success': False,
                    'error': 'Parent job has no nodes to edit'
                }), 400
            previous_content = json.dumps(nodes, indent=2)

        # Get source info for the job record
        source = source_index_service.get_source_from_index(project_id, source_id)
        if not source:
            return jsonify({
                'success': False,
                'error': f'Source not found: {source_id}'
            }), 404

        source_name = source.get('name', 'Unknown')

        # Create job record with edit lineage
        job_id = str(uuid.uuid4())
        studio_index_service.create_mind_map_job(
            project_id=project_id,
            job_id=job_id,
            source_id=source_id,
            source_name=source_name,
            direction=direction,
            parent_job_id=parent_job_id,
            edit_instructions=edit_instructions
        )

        # Submit background task with edit context
        task_service.submit_task(
            task_type="mind_map",
            target_id=job_id,
            callable_func=mind_map_service.generate_mind_map,
            project_id=project_id,
            source_id=source_id,
            job_id=job_id,
            direction=direction,
            previous_content=previous_content,
            edit_instructions=edit_instructions
        )

        return jsonify({
            'success': True,
            'job_id': job_id,
            'message': 'Mind map generation started',
            'source_name': source_name
        }), 202  # 202 Accepted - processing started

    except Exception as e:
        current_app.logger.error(f"Error starting mind map generation: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to start mind map generation: {str(e)}'
        }), 500


@studio_bp.route('/projects/<project_id>/studio/mind-map-jobs/<job_id>', methods=['GET'])
def get_mind_map_job_status(project_id: str, job_id: str):
    """
    Get the status of a mind map generation job.

    Response:
        - success: Boolean
        - job: Job record with status, progress, nodes (when ready)
    """
    try:
        job = studio_index_service.get_mind_map_job(project_id, job_id)

        if not job:
            return jsonify({
                'success': False,
                'error': f'Job not found: {job_id}'
            }), 404

        return jsonify({
            'success': True,
            'job': job
        }), 200

    except Exception as e:
        current_app.logger.error(f"Error getting mind map job status: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to get job status: {str(e)}'
        }), 500


@studio_bp.route('/projects/<project_id>/studio/mind-map-jobs', methods=['GET'])
def list_mind_map_jobs(project_id: str):
    """
    List all mind map jobs for a project.

    Query Parameters:
        - source_id: Optional filter by source

    Response:
        - success: Boolean
        - jobs: List of job records
    """
    try:
        source_id = request.args.get('source_id')
        jobs = studio_index_service.list_mind_map_jobs(project_id, source_id)

        # Filter out orphaned failed-edit jobs (error + parent_job_id).
        # These are leftover from edit failures and should never be shown.
        clean_jobs = [
            job for job in jobs
            if not (job.get("status") == "error" and job.get("parent_job_id"))
        ]

        return jsonify({
            'success': True,
            'jobs': clean_jobs,
            'count': len(clean_jobs)
        }), 200

    except Exception as e:
        current_app.logger.error(f"Error listing mind map jobs: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to list jobs: {str(e)}'
        }), 500


@studio_bp.route('/projects/<project_id>/studio/mind-map-jobs/<job_id>', methods=['DELETE'])
def delete_mind_map_job(project_id: str, job_id: str):
    """
    Delete a mind map job.

    Response:
        - success: Boolean
        - message: Status message
    """
    try:
        job = studio_index_service.get_mind_map_job(project_id, job_id)

        if not job:
            return jsonify({
                'success': False,
                'error': f'Mind map job {job_id} not found'
            }), 404

        studio_index_service.delete_mind_map_job(project_id, job_id)

        return jsonify({
            'success': True,
            'message': f'Mind map job {job_id} deleted'
        })

    except Exception as e:
        current_app.logger.error(f"Error deleting mind map job: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to delete mind map job: {str(e)}'
        }), 500
